include/graphblas/algorithms/amg/amg.hpp

```diff
@@ -0,0 +1,230 @@
+		 * @param[in] presmoother_steps number of pre-smoother steps, for multi-grid runs
+		 * @param[in] postsmoother_steps nomber of post-smoother steps, for multi-grid runs
+		 * @param[in] max_iterations maximum number if iterations the simulation may run for; once reached,
+		 *                           the simulation stops even if the residual norm is above \p tolerance
+		 * @param[in] tolerance the tolerance over the residual norm, i.e. the value of the residual norm to stop
+		 *                      the simulation at
+		 * @param[out] iterations numbers of iterations performed
+		 * @param[out] norm_residual norm of the final residual
+		 * @param[in] ring the ring to perform the operations on
+		 * @param[in] minus the \f$ - \f$ operator for vector subtractions
+		 * @return grb::RC::SUCCESS if the algorithm could correctly terminate, the error code of the first
+		 *                          unsuccessful operation otherwise
+		 */
+		template< typename IOType,
+			typename ResidualType,
+			typename NonzeroType,
+			typename InputType,
+			class Ring = Semiring< grb::operators::add< IOType >, grb::operators::mul< IOType >, grb::identities::zero, grb::identities::one >,
+			class Minus = operators::subtract< IOType > >
+		grb::RC amg( amg_data< IOType, NonzeroType, InputType > &data,
+			bool with_preconditioning,
+			const size_t presmoother_steps,
+			const size_t postsmoother_steps,
+			const size_t max_iterations,
+			const ResidualType tolerance,
+			size_t &iterations,
+			ResidualType &norm_residual,
+			const Ring &ring = Ring(),
+			const Minus &minus = Minus()
+		) {
+			ResidualType alpha;
+
+			const grb::Matrix< NonzeroType > &A { data.A };
+			grb::Vector< IOType > &x { data.x };
+			const grb::Vector< InputType > &b { data.b };
+			grb::Vector< IOType > &r { data.r };  // residual vector
+			grb::Vector< IOType > &p { data.p };  // direction vector
+			grb::Vector< IOType > &Ap { data.u }; // temp vector
+			grb::Vector< IOType > &z { data.z };  // pre-conditioned residual vector
+			grb::RC ret { SUCCESS };
+
+			ret = ret ? ret : grb::set( Ap, 0 );
+			ret = ret ? ret : grb::set( r, 0 );
+			ret = ret ? ret : grb::set( p, 0 );
+
+			ret = ret ? ret : grb::set( p, x );
+			ret = ret ? ret : grb::mxv( Ap, A, x, ring ); // Ap = A * x
+			assert( ret == SUCCESS );
+
+			ret = ret ? ret : grb::eWiseApply( r, b, Ap, minus ); // r = b - Ap;
+			assert( ret == SUCCESS );
+
+			norm_residual = ring.template getZero< ResidualType >();
+			ret = ret ? ret : grb::dot( norm_residual, r, r, ring ); // norm_residual = r' * r;
+			assert( ret == SUCCESS );
+
+			// compute sqrt to avoid underflow
+			norm_residual = std::sqrt( norm_residual );
+
+			// initial norm of residual
+			const ResidualType norm_residual_initial { norm_residual };
+			ResidualType old_r_dot_z { 0.0 }, r_dot_z { 0.0 }, beta { 0.0 };
+			size_t iter { 0 };
+
+#ifdef AMG_PRINT_STEPS
```

Comment:
Should this not be `_DEBUG` instead of `AMG_PRINT_STEPS` ? (Or maybe add
```
#ifdef _DEBUG
 #define AMG_PRINT_STEPS
#endif
```
in the header?